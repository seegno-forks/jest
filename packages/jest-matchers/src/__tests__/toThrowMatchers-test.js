/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @emails oncall+jsinfra
 */

'use strict';

const jestExpect = require('../').expect;
const skipOnWindows = require('skipOnWindows');

// Custom Error class because node versions have different stack trace strings.
class Error {
  constructor(message) {
    this.message = message;
    this.name = 'Error';
    this.stack =
      'Error\n' +
      '  at jestExpect' +
      ' (packages/jest-matchers/src/__tests__/toThrowMatchers-test.js:24:74)';
  }
}

['toThrowError', 'toThrow'].forEach(toThrow => {
  describe('.' + toThrow + '()', () => {
    skipOnWindows.suite();

    class Err extends Error {}
    class Err2 extends Error {}

    test('to throw or not to throw', () => {
      jestExpect(() => { throw new Error('apple'); })[toThrow]();
      jestExpect(() => {}).not[toThrow]();
    });

    describe('strings', () => {
      it('passes', () => {
        jestExpect(() => { throw new Error('apple'); })[toThrow]('apple');
        jestExpect(() => { throw new Error('banana'); })
          .not[toThrow]('apple');
        jestExpect(() => {}).not[toThrow]('apple');
        jestExpect(() => { throw new Error('apple'); })
          [toThrow]('apple', e => { expect(e).toBeInstanceOf(Error); });
        jestExpect(() => { throw new Error('banana'); })
          .not[toThrow]('apple', e => { expect(e).not.toBeInstanceOf(Error); });
      });

      test('did not throw at all', () => {
        expect(() => jestExpect(() => {})[toThrow]('apple'))
          .toThrowErrorMatchingSnapshot();
      });

      test('threw, but message did not match', () => {
        expect(() => {
          jestExpect(() => { throw new Error('apple'); })
            [toThrow]('banana');
        }).toThrowErrorMatchingSnapshot();
      });

      it('properly escapes strings when matching against errors', () => {
        jestExpect(() => { throw new TypeError('"this"? throws.'); })
          [toThrow]('"this"? throws.');
      });

      test('threw, but should not have', () => {
        expect(() => {
          jestExpect(() => { throw new Error('apple'); })
            .not[toThrow]('apple');
        }).toThrowErrorMatchingSnapshot();
      });

      test('threw, but should not have, and callback did not fail', () => {
        expect(() => {
          jestExpect(() => { throw new Error('apple'); })
            .not[toThrow]('apple', e => { expect(e).toBeInstanceOf(Error); });
        }).toThrowErrorMatchingSnapshot();
      });

      test('threw as expected, but callback failed', () => {
        expect(() => {
          jestExpect(() => { throw new Error('apple'); })
            [toThrow]('apple', e => { expect(e).not.toBeInstanceOf(Error); });
        }).toThrowErrorMatchingSnapshot();
      });
    });

    describe('regexp', () => {
      it('passes', () => {
        expect(() => { throw new Error('apple'); })[toThrow](/apple/);
        expect(() => { throw new Error('banana'); }).not[toThrow](/apple/);
        expect(() => {}).not[toThrow](/apple/);
        expect(() => { throw new Error('apple'); })
          [toThrow](/apple/, e => { expect(e).toBeInstanceOf(Error); });
        expect(() => { throw new Error('banana'); })
          .not[toThrow](/apple/, e => { expect(e).not.toBeInstanceOf(Error); });
      });

      test('did not throw at all', () => {
        expect(() => jestExpect(() => {})[toThrow](/apple/))
          .toThrowErrorMatchingSnapshot();
      });

      test('threw, but message did not match', () => {
        expect(() => {
          jestExpect(() => { throw new Error('apple'); })
            [toThrow](/banana/);
        }).toThrowErrorMatchingSnapshot();
      });

      test('threw, but should not have', () => {
        expect(() => {
          jestExpect(() => { throw new Error('apple'); })
            .not[toThrow](/apple/);
        }).toThrowErrorMatchingSnapshot();
      });

      test('threw, but should not have, and callback did not fail', () => {
        expect(() => {
          jestExpect(() => { throw new Error('apple'); })
            .not[toThrow]('apple', e => { expect(e).toBeInstanceOf(Error); });
        }).toThrowErrorMatchingSnapshot();
      });

      test('threw as expected, but callback failed', () => {
        expect(() => {
          jestExpect(() => { throw new Error('apple'); })
            [toThrow](/apple/, e => { expect(e).not.toBeInstanceOf(Error); });
        }).toThrowErrorMatchingSnapshot();
      });
    });

    describe('errors', () => {
      it('works', () => {
        it('passes', () => {
          jestExpect(() => { throw new Err(); })[toThrow](new Err());
          jestExpect(() => { throw new Err('Message'); })
            [toThrow](new Err('Message'));
          jestExpect(() => { throw new Err(); })[toThrow](new Error());
          jestExpect(() => { throw new Err(); }).not[toThrow](new Err2());
          jestExpect(() => {}).not[toThrow](new Err());
          jestExpect(() => { throw new Err(); })
            [toThrow](new Err(), e => { expect(e).toBeInstanceOf(Err); });
          jestExpect(() => { throw new Err(); })
            .not[toThrow](new Err2(), e => {
              expect(e).not.toBeInstanceOf(Err);
            });
        });
      });

      test('threw, but should not have, and callback did not fail', () => {
        expect(() => {
          jestExpect(() => { throw new Err(); })
            .not[toThrow](new Err(), e => { expect(e).toBeInstanceOf(Err); });
        }).toThrowErrorMatchingSnapshot();
      });

      test('threw as expected, but callback failed', () => {
        expect(() => {
          jestExpect(() => { throw new Err(); })[toThrow](new Err(), e => {
            expect(e).not.toBeInstanceOf(Err);
          });
        }).toThrowErrorMatchingSnapshot();
      });
    });

    describe('error class', () => {
      it('passes', () => {
        jestExpect(() => { throw new Err(); })[toThrow](Err);
        jestExpect(() => { throw new Err(); })[toThrow](Error);
        jestExpect(() => { throw new Err(); }).not[toThrow](Err2);
        jestExpect(() => {}).not[toThrow](Err);
        jestExpect(() => { throw new Err(); })[toThrow](Err, e => {
          expect(e).toBeInstanceOf(Err);
        });
        jestExpect(() => { throw new Err(); }).not[toThrow](Err2, e => {
          expect(e).not.toBeInstanceOf(Err);
        });
      });

      test('did not throw at all', () => {
        expect(() => expect(() => {})[toThrow](Err))
          .toThrowErrorMatchingSnapshot();
      });

      test('threw, but class did not match', () => {
        expect(() => {
          jestExpect(() => { throw new Err('apple'); })[toThrow](Err2);
        }).toThrowErrorMatchingSnapshot();
      });

      test('threw, but should not have', () => {
        expect(() => {
          jestExpect(() => { throw new Err('apple'); }).not[toThrow](Err);
        }).toThrowErrorMatchingSnapshot();
      });

      test('threw, but should not have, and callback did not fail', () => {
        expect(() => {
          jestExpect(() => { throw new Err(); })
            .not[toThrow](Err, e => { expect(e).toBeInstanceOf(Err); });
        }).toThrowErrorMatchingSnapshot();
      });

      test('threw as expected, but callback failed', () => {
        expect(() => {
          jestExpect(() => { throw new Err(); })[toThrow](Err, e => {
            expect(e).not.toBeInstanceOf(Err);
          });
        }).toThrowErrorMatchingSnapshot();
      });
    });

    test('invalid arguments', () => {
      expect(() => jestExpect(() => {})[toThrow](111))
        .toThrowErrorMatchingSnapshot();
    });
  });
});
