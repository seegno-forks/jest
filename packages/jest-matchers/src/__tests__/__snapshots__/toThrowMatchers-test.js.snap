exports[`.toThrow() error class did not throw at all 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).toThrow([22m[32mtype[39m[2m)[22m

Expected the function to throw an error of type:
  [32m\"Err\"[39m
But it didn\'t throw anything."
`;

exports[`.toThrow() error class threw as expected, but callback failed 1`] = `
"[2mexpect([22m[31mvalue[39m[2m).not.toBeInstanceOf([22m[32mconstructor[39m[2m)[22m

Expected value not to be an instance of:
  [32m\"Err\"[39m
Received:
  [31m{\"message\": undefined, \"name\": \"Error\", \"stack\": \"Error
  at jestExpect (packages/jest-matchers/src/__tests__/toThrowMatchers-test.js:24:74)\"}[39m
"
`;

exports[`.toThrow() error class threw, but class did not match 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).toThrow([22m[32mtype[39m[2m)[22m

Expected the function to throw an error of type:
  [32m\"Err2\"[39m
Instead, it threw:
[31m  Error      
      [2mat jestExpect ([22mpackages/jest-matchers/src/__tests__/toThrowMatchers-test.js[2m:24:74)[22m[39m"
`;

exports[`.toThrow() error class threw, but should not have 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrow([22m[32mtype[39m[2m)[22m

Expected the function not to throw an error of type:
  [32m\"Err\"[39m
Instead, it threw:
[31m  Error      
      [2mat jestExpect ([22mpackages/jest-matchers/src/__tests__/toThrowMatchers-test.js[2m:24:74)[22m[39m"
`;

exports[`.toThrow() error class threw, but should not have, and callback did not fail 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrow([22m[32mtype[39m[2m)[22m

Expected the function not to throw an error of type:
  [32m\"Err\"[39m
Instead, it threw:
[31m  Error      
      [2mat jestExpect ([22mpackages/jest-matchers/src/__tests__/toThrowMatchers-test.js[2m:24:74)[22m[39m"
`;

exports[`.toThrow() errors threw as expected, but callback failed 1`] = `
"[2mexpect([22m[31mvalue[39m[2m).not.toBeInstanceOf([22m[32mconstructor[39m[2m)[22m

Expected value not to be an instance of:
  [32m\"Err\"[39m
Received:
  [31m{\"message\": undefined, \"name\": \"Error\", \"stack\": \"Error
  at jestExpect (packages/jest-matchers/src/__tests__/toThrowMatchers-test.js:24:74)\"}[39m
"
`;

exports[`.toThrow() errors threw, but should not have, and callback did not fail 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrow([22m[32merror[39m[2m)[22m

Expected the function not to throw an error matching:
  [32m{\"message\": undefined, \"name\": \"Error\", \"stack\": \"Error
  at jestExpect (packages/jest-matchers/src/__tests__/toThrowMatchers-test.js:24:74)\"}[39m
Instead, it threw:
[31m  Error      
      [2mat jestExpect ([22mpackages/jest-matchers/src/__tests__/toThrowMatchers-test.js[2m:24:74)[22m[39m"
`;

exports[`.toThrow() invalid arguments 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrow([22m[32mnumber[39m[2m)[22m

Unexpected argument passed.
Expected: [32m\"string\"[39m, [32m\"Error (type)\"[39m or [32m\"regexp\"[39m.
Got:
  string: [32m\"111\"[39m"
`;

exports[`.toThrow() regexp did not throw at all 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).toThrow([22m[32mregexp[39m[2m)[22m

Expected the function to throw an error matching:
  [32m/apple/[39m
But it didn\'t throw anything."
`;

exports[`.toThrow() regexp threw as expected, but callback failed 1`] = `
"[2mexpect([22m[31mvalue[39m[2m).not.toBeInstanceOf([22m[32mconstructor[39m[2m)[22m

Expected value not to be an instance of:
  [32m\"Error\"[39m
Received:
  [31m{\"message\": \"apple\", \"name\": \"Error\", \"stack\": \"Error
  at jestExpect (packages/jest-matchers/src/__tests__/toThrowMatchers-test.js:24:74)\"}[39m
"
`;

exports[`.toThrow() regexp threw, but message did not match 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).toThrow([22m[32mregexp[39m[2m)[22m

Expected the function to throw an error matching:
  [32m/banana/[39m
Instead, it threw:
[31m  Error      
      [2mat jestExpect ([22mpackages/jest-matchers/src/__tests__/toThrowMatchers-test.js[2m:24:74)[22m[39m"
`;

exports[`.toThrow() regexp threw, but should not have 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrow([22m[32mregexp[39m[2m)[22m

Expected the function not to throw an error matching:
  [32m/apple/[39m
Instead, it threw:
[31m  Error      
      [2mat jestExpect ([22mpackages/jest-matchers/src/__tests__/toThrowMatchers-test.js[2m:24:74)[22m[39m"
`;

exports[`.toThrow() regexp threw, but should not have, and callback did not fail 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrow([22m[32mstring[39m[2m)[22m

Expected the function not to throw an error matching:
  [32m\"apple\"[39m
Instead, it threw:
[31m  Error      
      [2mat jestExpect ([22mpackages/jest-matchers/src/__tests__/toThrowMatchers-test.js[2m:24:74)[22m[39m"
`;

exports[`.toThrow() strings did not throw at all 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).toThrow([22m[32mstring[39m[2m)[22m

Expected the function to throw an error matching:
  [32m\"apple\"[39m
But it didn\'t throw anything."
`;

exports[`.toThrow() strings threw as expected, but callback failed 1`] = `
"[2mexpect([22m[31mvalue[39m[2m).not.toBeInstanceOf([22m[32mconstructor[39m[2m)[22m

Expected value not to be an instance of:
  [32m\"Error\"[39m
Received:
  [31m{\"message\": \"apple\", \"name\": \"Error\", \"stack\": \"Error
  at jestExpect (packages/jest-matchers/src/__tests__/toThrowMatchers-test.js:24:74)\"}[39m
"
`;

exports[`.toThrow() strings threw, but message did not match 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).toThrow([22m[32mstring[39m[2m)[22m

Expected the function to throw an error matching:
  [32m\"banana\"[39m
Instead, it threw:
[31m  Error      
      [2mat jestExpect ([22mpackages/jest-matchers/src/__tests__/toThrowMatchers-test.js[2m:24:74)[22m[39m"
`;

exports[`.toThrow() strings threw, but should not have 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrow([22m[32mstring[39m[2m)[22m

Expected the function not to throw an error matching:
  [32m\"apple\"[39m
Instead, it threw:
[31m  Error      
      [2mat jestExpect ([22mpackages/jest-matchers/src/__tests__/toThrowMatchers-test.js[2m:24:74)[22m[39m"
`;

exports[`.toThrow() strings threw, but should not have, and callback did not fail 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrow([22m[32mstring[39m[2m)[22m

Expected the function not to throw an error matching:
  [32m\"apple\"[39m
Instead, it threw:
[31m  Error      
      [2mat jestExpect ([22mpackages/jest-matchers/src/__tests__/toThrowMatchers-test.js[2m:24:74)[22m[39m"
`;

exports[`.toThrowError() error class did not throw at all 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).toThrowError([22m[32mtype[39m[2m)[22m

Expected the function to throw an error of type:
  [32m\"Err\"[39m
But it didn\'t throw anything."
`;

exports[`.toThrowError() error class threw as expected, but callback failed 1`] = `
"[2mexpect([22m[31mvalue[39m[2m).not.toBeInstanceOf([22m[32mconstructor[39m[2m)[22m

Expected value not to be an instance of:
  [32m\"Err\"[39m
Received:
  [31m{\"message\": undefined, \"name\": \"Error\", \"stack\": \"Error
  at jestExpect (packages/jest-matchers/src/__tests__/toThrowMatchers-test.js:24:74)\"}[39m
"
`;

exports[`.toThrowError() error class threw, but class did not match 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).toThrowError([22m[32mtype[39m[2m)[22m

Expected the function to throw an error of type:
  [32m\"Err2\"[39m
Instead, it threw:
[31m  Error      
      [2mat jestExpect ([22mpackages/jest-matchers/src/__tests__/toThrowMatchers-test.js[2m:24:74)[22m[39m"
`;

exports[`.toThrowError() error class threw, but should not have 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowError([22m[32mtype[39m[2m)[22m

Expected the function not to throw an error of type:
  [32m\"Err\"[39m
Instead, it threw:
[31m  Error      
      [2mat jestExpect ([22mpackages/jest-matchers/src/__tests__/toThrowMatchers-test.js[2m:24:74)[22m[39m"
`;

exports[`.toThrowError() error class threw, but should not have, and callback did not fail 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowError([22m[32mtype[39m[2m)[22m

Expected the function not to throw an error of type:
  [32m\"Err\"[39m
Instead, it threw:
[31m  Error      
      [2mat jestExpect ([22mpackages/jest-matchers/src/__tests__/toThrowMatchers-test.js[2m:24:74)[22m[39m"
`;

exports[`.toThrowError() errors threw as expected, but callback failed 1`] = `
"[2mexpect([22m[31mvalue[39m[2m).not.toBeInstanceOf([22m[32mconstructor[39m[2m)[22m

Expected value not to be an instance of:
  [32m\"Err\"[39m
Received:
  [31m{\"message\": undefined, \"name\": \"Error\", \"stack\": \"Error
  at jestExpect (packages/jest-matchers/src/__tests__/toThrowMatchers-test.js:24:74)\"}[39m
"
`;

exports[`.toThrowError() errors threw, but should not have, and callback did not fail 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowError([22m[32merror[39m[2m)[22m

Expected the function not to throw an error matching:
  [32m{\"message\": undefined, \"name\": \"Error\", \"stack\": \"Error
  at jestExpect (packages/jest-matchers/src/__tests__/toThrowMatchers-test.js:24:74)\"}[39m
Instead, it threw:
[31m  Error      
      [2mat jestExpect ([22mpackages/jest-matchers/src/__tests__/toThrowMatchers-test.js[2m:24:74)[22m[39m"
`;

exports[`.toThrowError() invalid arguments 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowError([22m[32mnumber[39m[2m)[22m

Unexpected argument passed.
Expected: [32m\"string\"[39m, [32m\"Error (type)\"[39m or [32m\"regexp\"[39m.
Got:
  string: [32m\"111\"[39m"
`;

exports[`.toThrowError() regexp did not throw at all 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).toThrowError([22m[32mregexp[39m[2m)[22m

Expected the function to throw an error matching:
  [32m/apple/[39m
But it didn\'t throw anything."
`;

exports[`.toThrowError() regexp threw as expected, but callback failed 1`] = `
"[2mexpect([22m[31mvalue[39m[2m).not.toBeInstanceOf([22m[32mconstructor[39m[2m)[22m

Expected value not to be an instance of:
  [32m\"Error\"[39m
Received:
  [31m{\"message\": \"apple\", \"name\": \"Error\", \"stack\": \"Error
  at jestExpect (packages/jest-matchers/src/__tests__/toThrowMatchers-test.js:24:74)\"}[39m
"
`;

exports[`.toThrowError() regexp threw, but message did not match 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).toThrowError([22m[32mregexp[39m[2m)[22m

Expected the function to throw an error matching:
  [32m/banana/[39m
Instead, it threw:
[31m  Error      
      [2mat jestExpect ([22mpackages/jest-matchers/src/__tests__/toThrowMatchers-test.js[2m:24:74)[22m[39m"
`;

exports[`.toThrowError() regexp threw, but should not have 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowError([22m[32mregexp[39m[2m)[22m

Expected the function not to throw an error matching:
  [32m/apple/[39m
Instead, it threw:
[31m  Error      
      [2mat jestExpect ([22mpackages/jest-matchers/src/__tests__/toThrowMatchers-test.js[2m:24:74)[22m[39m"
`;

exports[`.toThrowError() regexp threw, but should not have, and callback did not fail 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowError([22m[32mstring[39m[2m)[22m

Expected the function not to throw an error matching:
  [32m\"apple\"[39m
Instead, it threw:
[31m  Error      
      [2mat jestExpect ([22mpackages/jest-matchers/src/__tests__/toThrowMatchers-test.js[2m:24:74)[22m[39m"
`;

exports[`.toThrowError() strings did not throw at all 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).toThrowError([22m[32mstring[39m[2m)[22m

Expected the function to throw an error matching:
  [32m\"apple\"[39m
But it didn\'t throw anything."
`;

exports[`.toThrowError() strings threw as expected, but callback failed 1`] = `
"[2mexpect([22m[31mvalue[39m[2m).not.toBeInstanceOf([22m[32mconstructor[39m[2m)[22m

Expected value not to be an instance of:
  [32m\"Error\"[39m
Received:
  [31m{\"message\": \"apple\", \"name\": \"Error\", \"stack\": \"Error
  at jestExpect (packages/jest-matchers/src/__tests__/toThrowMatchers-test.js:24:74)\"}[39m
"
`;

exports[`.toThrowError() strings threw, but message did not match 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).toThrowError([22m[32mstring[39m[2m)[22m

Expected the function to throw an error matching:
  [32m\"banana\"[39m
Instead, it threw:
[31m  Error      
      [2mat jestExpect ([22mpackages/jest-matchers/src/__tests__/toThrowMatchers-test.js[2m:24:74)[22m[39m"
`;

exports[`.toThrowError() strings threw, but should not have 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowError([22m[32mstring[39m[2m)[22m

Expected the function not to throw an error matching:
  [32m\"apple\"[39m
Instead, it threw:
[31m  Error      
      [2mat jestExpect ([22mpackages/jest-matchers/src/__tests__/toThrowMatchers-test.js[2m:24:74)[22m[39m"
`;

exports[`.toThrowError() strings threw, but should not have, and callback did not fail 1`] = `
"[2mexpect([22m[31mfunction[39m[2m).not.toThrowError([22m[32mstring[39m[2m)[22m

Expected the function not to throw an error matching:
  [32m\"apple\"[39m
Instead, it threw:
[31m  Error      
      [2mat jestExpect ([22mpackages/jest-matchers/src/__tests__/toThrowMatchers-test.js[2m:24:74)[22m[39m"
`;
